### PR TITLE
add support for --update-lmod-caches (disabled by default) and --compile-lmod-caches (WIP)

### PR DESCRIPTION
**Pull request by boegel**
@rtmclay: it would be great if you could give this a review, especially w.r.t. the FIXME I have in there

### CODE REVIEW

File: easybuild/tools/modules.py [hunk 1]
```diff
@@ -831,33 +834,54 @@ def available(self, mod_name=None):
 
         return correct_real_mods
 
+    def compile_cache(self, cache_fp):
+        """Compile Lmod cache file using luac."""
+        if build_option('compile_lmod_caches'):
+            # FIXME: ask Lmod to compile, to ensure that luac that matches the lua used by Lmod is used?
```
Comment (stdweird) [inline easybuild/tools/modules.py]: :+1: on letting lmod compile the cache. then we also don't need to test if luac is in PATH etc etc.

File: easybuild/tools/modules.py [hunk 2]
```diff
@@ -831,33 +834,54 @@ def available(self, mod_name=None):
...
-                cache_file.write(stdout)
-                cache_file.close()
-            except (IOError, OSError), err:
-                self.log.error("Failed to update Lmod spider cache %s: %s" % (cache_filefn, err))
+                # run spider to create cache contents
+                spider_cmd = os.path.join(os.path.dirname(self.cmd), 'spider')
+                cmd = [spider_cmd, '-o', cache_type, os.environ['MODULEPATH']]
+                self.log.debug("Running command '%s'..." % ' '.join(cmd))
+
+                # we can't use run_cmd, since we need to separate stdout from stderr
```
Comment (stdweird) [inline easybuild/tools/modules.py]: what's so special about stderr? there's no proper exitcode?

File: easybuild/tools/modules.py [hunk 3]
```diff
@@ -831,33 +834,54 @@ def available(self, mod_name=None):
...
+                else:
+                    cache_fp = os.path.join(cache_dir, '%s.lua' % cache_type)
+                    self.log.debug("Updating Lmod spider cache %s with output from '%s'" % (cache_fp, ' '.join(cmd)))
+
+                    # back existing cache by copying it to <cache>T.old.lua
+                    # Lmod also considers this filename for the cache in case <cache>T.lua isn't found
+                    if os.path.exists(cache_fp):
+                        old_cache_fp = '%s.old.lua' % os.path.splitext(cache_fp)[0]
+                        copy_file(cache_fp, old_cache_fp)
+                        self.compile_cache(old_cache_fp)
```
Comment (stdweird) [inline easybuild/tools/modules.py]: why recompile and not move the old cache (if available)?

File: easybuild/tools/modules.py [hunk 4]
```diff
@@ -831,33 +834,54 @@ def available(self, mod_name=None):
...
-                cache_dir = os.path.dirname(cache_filefn)
-                if not os.path.exists(cache_dir):
-                    mkdir(cache_dir, parents=True)
-                cache_file = open(cache_filefn, 'w')
-                cache_file.write(stdout)
-                cache_file.close()
-            except (IOError, OSError), err:
-                self.log.error("Failed to update Lmod spider cache %s: %s" % (cache_filefn, err))
+                # run spider to create cache contents
+                spider_cmd = os.path.join(os.path.dirname(self.cmd), 'spider')
```
Comment (stdweird) [inline easybuild/tools/modules.py]: why not use https://github.com/TACC/Lmod/blob/master/contrib/BuildSystemCacheFile/createSystemCacheFile.sh? isn't it part of a lmod distribution?

File: easybuild/tools/options.py [hunk 5]
```diff
@@ -177,6 +178,7 @@ def override_options(self):
             'allow-modules-tool-mismatch': ("Allow mismatch of modules tool and definition of 'module' function",
                                             None, 'store_true', False),
             'cleanup-builddir': ("Cleanup build dir after successful installation.", None, 'store_true', True),
+            'compile-lmod-caches': ("Compile Lmod cache files with luac after updating them", None, 'store_true', False),
```
Comment (stdweird) [inline easybuild/tools/options.py]: i don't like the lmod option in the name. there are other module tools that can have caches

Comment (stdweird) [inline easybuild/tools/options.py]: what does this do? should be just a toggle, that if you run `--update-cache` also compiles it (but i don't like the idea. lmod should do what is best, i.e. compile if luac is avail).

File: easybuild/tools/options.py [hunk 6]
```diff
@@ -197,10 +199,12 @@ def override_options(self):
             'set-gid-bit': ("Set group ID bit on newly created directories", None, 'store_true', False),
             'sticky-bit': ("Set sticky bit on newly created directories", None, 'store_true', False),
             'skip-test-cases': ("Skip running test cases", None, 'store_true', False, 't'),
-            'umask': ("umask to use (e.g. '022'); non-user write permissions on install directories are removed",
-                      None, 'store', None),
             'optarch': ("Set architecture optimization, overriding native architecture optimizations",
                         None, 'store', None),
+            'umask': ("umask to use (e.g. '022'); non-user write permissions on install directories are removed",
+                      None, 'store', None),
+            'update-lmod-caches': ("Update Lmod cache files in specified directory", None, 'store_or_None',
```
Comment (stdweird) [inline easybuild/tools/options.py]: should be module tool flavour neutral.

File: easybuild/tools/options.py [hunk 7]
```diff
@@ -405,6 +409,11 @@ def postprocess(self):
             if token is None:
                 self.log.error("Failed to obtain required GitHub token for user '%s'" % self.options.github_user)
 
+        # options w.r.t. updating Lmod cache only make sense when Lmod is selected modules tool
+        if self.options.update_lmod_caches:
+            if self.options.modules_tool != Lmod.__name__:
```
Comment (stdweird) [inline easybuild/tools/options.py]: should be a call to the modules tool class like `klass.can_create_cache` or something like that

File: easybuild/tools/modules.py (hunk 1 above)
Comment (boegel) [inline easybuild/tools/modules.py]: that's just a side-effect, but yes, that's good

it's important that Lmod uses the `luac` that matches the `lua` it's using, since the Lua version has to be part of the filename of the compiled cache (I still need to fix that here)

File: easybuild/tools/modules.py (hunk 2 above)
Comment (boegel) [inline easybuild/tools/modules.py]: `spider` acts like `lmod` and `modulecmd`, I think: the cache contents to `stdout`, any messages (also non-error) to `stderr`

File: easybuild/tools/modules.py (hunk 3 above)
Comment (boegel) [inline easybuild/tools/modules.py]: Lmod also picks up `moduleT.old.luac_<luaver>` and `moduleT.old.lua`, if they're there, to cover the rare case where `moduleT.lua` is not in place yet, but the previous version is still around

File: easybuild/tools/modules.py (hunk 4 above)
Comment (boegel) [inline easybuild/tools/modules.py]: I don't think it gets installed as a part of Lmod, and it needs to be tweaked anyway, it has hardcoded stuff in it specific to TACC.

I'm just redoing what that script implements, with proper logging (coming up) and error handling in Python, etc.

File: easybuild/tools/options.py (hunk 5 above)
Comment (boegel) [inline easybuild/tools/options.py]: `spider` just dumps the cache to `stdout`, there is no tool included in Lmod (yet?) that takes care of the `.old`, compiling, etc. I'm making the compilation optional here, since `luac` may not be around, and for older Lmod version the compiled cache is useless (since Lmod doesn't pick it up).

File: easybuild/tools/options.py (hunk 6 above)
Comment (boegel) [inline easybuild/tools/options.py]: there is no caching in Tcl(/C) env mods... suggestion for a better name?

at least that makes it clear from the name of the command line option that this is specific to Lmod

File: easybuild/tools/options.py (hunk 7 above)
Comment (boegel) [inline easybuild/tools/options.py]: hmm, good suggestion

or maybe `is_lmod()` ;-)

**Comment by pforai** [comment]
Can we look into drop using $LMOD_IGNORE_CACHE=1 globally?

File: easybuild/tools/modules.py [hunk 8]
```diff
@@ -48,12 +48,15 @@
...
 from easybuild.tools.environment import restore_env
-from easybuild.tools.filetools import convert_name, mkdir, read_file, path_matches, which
+from easybuild.tools.filetools import convert_name, copy_file, mkdir, read_file, path_matches, which, write_file
 from easybuild.tools.module_naming_scheme import DEVEL_MODULE_SUFFIX
 from easybuild.tools.run import run_cmd
 from easybuild.tools.toolchain import DUMMY_TOOLCHAIN_NAME, DUMMY_TOOLCHAIN_VERSION
 from vsc.utils.missing import nub
 
+# relative to user's home directory
+LMOD_USER_CACHE_RELDIR = '.lmod.d/.cache'
```
Comment (wpoely86) [inline easybuild/tools/modules.py]:
```
  --with-useDotFiles=yes/no
                          If yes use ~/.lmod.d/.cache, if no use
                          ~/.lmod.d/__cache__
```

File: easybuild/tools/modules.py (hunk 2 above)
Comment (wpoely86) [inline easybuild/tools/modules.py]: Shouldn't we then extand `run_cmd` for this use case?

Comment (boegel) [inline easybuild/tools/modules.py]: we should start using the `run` commands that are provided by `vsc.utils.run`, and the function in there should be enhanced to allow spltiting stdout/stderr

that's a long-term effort though, since either we redefine `run_cmd` to use those, or we drop/deprecate `run_cmd` all together

File: easybuild/tools/modules.py (hunk 3 above)
Comment (boegel) [inline easybuild/tools/modules.py]: ah, yes, but you're right, I should just move the old cache, rather than recompile is, just realised what you intended to point out

File: easybuild/tools/modules.py (hunk 8 above)
Comment (boegel) [inline easybuild/tools/modules.py]: this is just a default, `--update-lmod-caches` takes an argument if you want to update caches in another location

since `.lmod.d/.cache` is the default in Lmod, I think just keeping this as is makes sense

File: easybuild/tools/options.py (hunk 7 above)
Comment (boegel) [inline easybuild/tools/options.py]: added a static method `is_lmod` in `ModulesTool`

**Comment by boegel** [comment]
@pforai: OK, I'll try and tackle that bit in here, as discussed... It should boil down to always updating the cache before every build is started. Although this will take a little while, it will avoid having to scan the module tree multiple times when running, since `module avail`, which is done for each dependency, etc., will become very fast...

**Comment by boegel** [comment]
@pforai: I took a look, and there's more needed to make sure we can not ignore the cache...

We need to make sure that we can/should update the cache actively being used by Lmod in the environment we're running in. This could be a system cache (where we may not have permissions to to update it), or just our local user cache.

So, it seems like we need the following cmdline options:

* `--ignore-lmod-caches` (enabled by default, to make sure?)
* `--update-lmod-caches[=<dir>` (which is there now)

Iff `--ignore-lmod-caches` is set to `False`, we need to somehow query Lmod which cache it will be using, an try to update it. If that works, we can continue. If not, we need to scream bloody murder and end it before it goes bad somehow (e.g. accidentally reinstalling something because we don't see it in `ml av`).

Auto-updating the 'live' Lmod cache is probably also not a good idea in a general sense, since we don't know for which `$MODULEPATH` it was built... Although it may be fine to assume that the `$MODULEPATH` is correct when `eb` is being asked to update the `live` cache.

In short, I need to give this more thought, so we should probably stay away from *also* tackling that in this PR.

**Comment by boegel** [comment]
@stdweird: please re-review?

**Comment by boegel** [comment]
@pforai: One thing we can do is generate our own Lmod cache in `/tmp`, and then configure Lmod to use that cache via `$LMOD_RC`. With that in place, we can disable `LMOD_IGNORE_CACHE`...

**Comment by rtmclay** [comment]
This looks O.K. to me.  You might want to add support for -D with the spider command as a way to track down Lmod problems when building the spider cache.

File: easybuild/tools/options.py [hunk 9]
```diff
@@ -197,10 +198,13 @@ def override_options(self):
             'set-gid-bit': ("Set group ID bit on newly created directories", None, 'store_true', False),
             'sticky-bit': ("Set sticky bit on newly created directories", None, 'store_true', False),
             'skip-test-cases': ("Skip running test cases", None, 'store_true', False, 't'),
-            'umask': ("umask to use (e.g. '022'); non-user write permissions on install directories are removed",
-                      None, 'store', None),
             'optarch': ("Set architecture optimization, overriding native architecture optimizations",
                         None, 'store', None),
+            'umask': ("umask to use (e.g. '022'); non-user write permissions on install directories are removed",
+                      None, 'store', None),
+            'update-lmod-cache': ("Update Lmod cache files in specified dir after generating module file",
```
Comment (stdweird) [inline easybuild/tools/options.py]: (again) module tool neutral option name

File: easybuild/tools/modules.py [hunk 10]
```diff
@@ -832,32 +874,72 @@ def available(self, mod_name=None):
...
-                cache_dir = os.path.dirname(cache_filefn)
-                if not os.path.exists(cache_dir):
-                    mkdir(cache_dir, parents=True)
-                cache_file = open(cache_filefn, 'w')
-                cache_file.write(stdout)
-                cache_file.close()
-            except (IOError, OSError), err:
-                self.log.error("Failed to update Lmod spider cache %s: %s" % (cache_filefn, err))
+        """Update Lmod cache after new modules were added."""
+        cache_dir = None
```
Comment (stdweird) [inline easybuild/tools/modules.py]: again, why is this not simply calling https://github.com/TACC/Lmod/blob/master/contrib/BuildSystemCacheFile/createSystemCacheFile.sh
@rtmclay is the createSystemCacheFile.sh not part of the supported lmod release?

File: easybuild/tools/modules.py [hunk 11]
```diff
@@ -137,6 +140,11 @@ class ModulesTool(object):
 
     __metaclass__ = Singleton
 
+    @staticmethod
+    def is_lmod():
```
Comment (stdweird) [inline easybuild/tools/modules.py]: neutral names pls

File: easybuild/tools/options.py [hunk 12]
```diff
@@ -405,6 +409,11 @@ def postprocess(self):
             if token is None:
                 self.log.error("Failed to obtain required GitHub token for user '%s'" % self.options.github_user)
 
+        # options w.r.t. updating Lmod cache only make sense when Lmod is selected modules tool
+        modules_tool_class = modules_tool(name=self.options.modules_tool, return_class=True)
+        if self.options.update_lmod_cache and not modules_tool_class.is_lmod():
+            self.log.error("Options related to Lmod cache are only supported if Lmod is used as a modules tool.")
```
Comment (stdweird) [inline easybuild/tools/options.py]: there shouldn't be any options related to only one modules tool. ever.

**Comment by rtmclay** [comment]
the script createSystemCacheFile.sh is part of the standard Lmod distribution.  But it has to be modified to be used at each site.

File: easybuild/tools/modules.py (hunk 10 above)
Comment (wpoely86) [inline easybuild/tools/modules.py]: because `createSystemCacheFile.sh` does not get installed and it needs manual tweaking before you can use it. It's an example how to create the spider cache (hence why it is in the `contrib` directory).

**Comment by boegel** [comment]
@rtmclay: is there any reason why there is no dedicated option in Lmod to update the cache(s) at a specific location?

**Comment by stdweird** [comment]
@rtmclay what needs customisation? it would be better if EB can just use it instead trying to reproduce it in python (it would be even better if lmod had a `lmod --make-cache` option).

File: easybuild/tools/options.py (hunk 9 above)
Comment (boegel) [inline easybuild/tools/options.py]: OK, I'll make this more neutral, something like `update-modules-tool-cache`.

**Comment by boegel** [comment]
@stdweird: the location of the cache and timestamp files, potentially also the `$MODULEPATH` for which the cache is built, the `luac` being used (which has to match the one Lmod is using), etc.

But, I agree that there should be a dedicated cmdline option in Lmod for doing this. The only reason I'm reproducing it now is because there isn't (yet). If there was, I would use it for sure.

File: easybuild/tools/modules.py (hunk 10 above)
Comment (stdweird) [inline easybuild/tools/modules.py]: @wpoely86 why wouldn't it get installed? sure EB can do that (and package maintainers should ship it too)
what needs custiomisation? is it only `LMOD_DIR`?

**Comment by rtmclay** [comment]
The script createSystemCacheFile.sh can be modified to take arguments but having the lmod command do it is just wrong.   I'm not going to take bug reports that updating the system cache fails when the user doesn't have permission to write to that directory.   I can modify it when I get back.  Or I'll take a PR on that.

File: easybuild/tools/modules.py (hunk 10 above)
Comment (boegel) [inline easybuild/tools/modules.py]: `createSystemCacheFile.sh` doesn't solve our problem.

a dedicated cmdline option in Lmod would be great though, so let's see if that can works out

This PR can be on hold until it's there (or until we get final word it won't happen), it's not going in for EB v2.0 anyway (needs more testing using `--only-module`, which is not ready yet).

**Comment by boegel** [comment]
Jenkins: test this please

**Comment by boegel** [comment]
Jenkins: test this please

**Comment by akesandgren** [comment]
@boegel Relevance?
